crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,239 @@
+
+    /// The displayed lore of the associated `ItemStack`.
+    pub lore: String,
+
+    /// The damage taken by the `ItemStack`.
+    pub damage: Option<u32>,
+
+    /// The cost of repairing the `ItemStack`.
+    pub repair_cost: Option<u32>,
+
+    /// The enchantments applied to this `ItemStack`.
+    pub enchantments: Vec<Enchantment>,
+}
+
+impl ItemStack {
+    /// Creates a new `ItemStack` with the default name (title)
+    /// no lore, no damage, no repair cost and no enchantments.
+    pub fn new(item: Item, count: u32) -> Self {
+        Self {
+            item,
+            count,
+            meta: Some(ItemStackMeta {
+                title: String::from(item.name()),
+                lore: "".to_string(),
+                damage: None,
+                repair_cost: None,
+                enchantments: vec![],
+            }),
+        }
+    }
+
+    /// Returns whether the given item stack has
+    /// the same type as (but not necessarily the same
+    /// amount as) `self`.
+    pub fn has_same_type(&self, other: &Self) -> bool {
+        self.item == other.item
+    }
+
+    /// Returns whether the given item stack has the same damage
+    /// as `self`.
+    pub fn has_same_damage(&self, other: &Self) -> bool {
+        if let (Some(self_meta), Some(other_meta)) = (self.meta.clone(), other.meta.clone()) {
+            self_meta.damage == other_meta.damage
+        } else {
+            self.meta.is_none() && other.meta.is_none()
+        }
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same count as (but not necessarily the same
+    /// type as) `self`.
+    pub fn has_same_count(&self, other: &Self) -> bool {
+        self.count == other.count
+    }
+
+    /// Returns whether the given `ItemStack` has the same
+    /// type and count as `self`.
+    pub fn has_same_type_and_count(&self, other: &Self) -> bool {
+        self.item == other.item && self.count == other.count
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same type and damage as `self`.
+    pub fn has_same_type_and_damage(&self, other: &Self) -> bool {
+        self.item == other.item && self.has_same_damage(other)
+    }
+
+    /// Returns the item type for this `ItemStack`.
+    pub fn item(&self) -> Item {
+        self.item
+    }
+
+    /// Returns the number of items in this `ItemStack`.
+    pub fn count(&self) -> u32 {
+        self.count
+    }
+
+    /// Adds more items to this `ItemStack`. Returns the new count.
+    pub fn add(&mut self, count: u32) -> Result<u32, ()> {
+        if self.count + count > self.item.stack_size() {
+            return Err(());
+        }
+        self.count += count;
+        Ok(self.count)
+    }
+
+    /// Adds more items to this `ItemStack`. Does not check if the
+    /// addition will make the count to be greater than the
+    /// stack size. Returns the new count.
+    pub fn unchecked_add(&mut self, count: u32) -> u32 {
+        self.count += count;
```

Comment:
Since we are using a `u32` and not `i32` we could overflow into the negatives on the client. We should maybe limit it to `2^31-1`?

Reply:
I believe the unchecked function should be provided AS IS without checks of any kind.